src/server/db_slice.cc

```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
   events_.update = 0;
 }
 
+void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice& keys) {
+  if (conn.IsExpired()) {
+    DVLOG(2) << "Connection expired, exiting TrackKey function.";
+    return;
+  }
+
+  DVLOG(2) << "Start tracking keys for client ID: " << conn.GetClientId()
+           << " with thread ID: " << conn.Thread();
+  for (auto key : keys) {
+    DVLOG(2) << "Inserting client ID " << conn.GetClientId()
+             << " into the tracking client set of key " << key;
+    client_tracking_map_[key].insert(conn);
+  }
+}
+
+void DbSlice::SendInvalidationTrackingMessage(const std::string_view& key) {
+  if (client_tracking_map_.find(key) != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = client_tracking_map_[key];
+    DVLOG(2) << "Garbage collect clients that are no longer tracking... ";
+    auto is_closed_or_not_tracking = [](const facade::Connection::WeakRef& p) {
+      return (p.IsExpired() || (!p.Get()->IsTrackingOn()));
```

Comment:
Technically, this Get here is unsafe. In practice the the other thread can never be so fast to delete the connection faster than reading a variable...

Better do it from cb, as you drop them below either way